src/utils/polygon.cpp

```diff
@@ -370,21 +399,58 @@ void PolygonRef::simplify(const coord_t smallest_line_segment_squared, const coo
         //h^2 = L^2 / b^2     [factor the divisor]
         const coord_t height_2 = area_removed_so_far * area_removed_so_far / base_length_2;
         if ((height_2 <= 1 //Almost exactly colinear (barring rounding errors).
-            && LinearAlg2D::getDist2FromLine(current, previous, next) <= 1) // make sure that height_2 is not small because of cancellation of positive and negative areas
-            || (length2 < smallest_line_segment_squared
-                && next_length2 < smallest_line_segment_squared // Segments are small
-                && height_2 <= allowed_error_distance_squared) // removing the vertex doesn't introduce too much error.
-        )
+            && LinearAlg2D::getDist2FromLine(current, previous, next) <= 1)) // make sure that height_2 is not small because of cancellation of positive and negative areas
         {
-            continue; //Remove the vertex.
+            continue;
+        }
+  
+        if (length2 < smallest_line_segment_squared
+            && height_2 <= allowed_error_distance_squared) // removing the vertex doesn't introduce too much error.)
+        {
+            if (next_length2 > smallest_line_segment_squared)
+            {
+                // Special case; The next line is long. If we were to remove this, it could happen that we get quite noticeable artifacts.
+                // We should instead move this point to a location where both edges are kept and then remove the previous point that we wanted to keep.
+                // By taking the intersection of these two lines, we get a point that perseves the direction (so it makes the corner a bit more pointy)
+                Point intersection_point;
+                bool has_intersection = lineLineIntersection(previous_previous, previous, current, next, &intersection_point);
+                
+                if (!has_intersection || LinearAlg2D::getDist2FromLine(intersection_point, previous, current) > allowed_error_distance_squared)
+                {
+                    if(length2 < 25)
+                    {
+                        // We're allowed to always delete segments of less than 5 micron.
```

Comment:
Wait no this is dangerous! The whole idea of this function was to avoid escalating segment removal. The default argument `allowed_error_distance` is already 5mu; this whole function was built to be able to remove 5mu segments, but do it in a safe way.

The continue here should be removed.

Reply:
It only does it if the intersection fails. The intersection only fails if the direction of the two lines is almost exactly the same. 

So it's dangerous if you don't take it into context.

Reply:
The intersection only fails if the direction of the previous and next line is almost exactly the same, but not the same as the current line segment, so it's not like the collinearity means that this case can be ignored.

> The alternative is that I simply always delete the point, regardless if the length is <5mu, which is the solution I had before this.

The alternative is option C or E, not option B. That is: either leave the small segment in, or move either of the verts of the current segment along the long segment, such that the current segment becomes as long as `smallest_line_segment`.

> If we don't delete these tiny points, we still get these very tiny line segments in some spots, which wreak havoc on a print.

That may be the case for some printers, but I'd like to challenge the idea that this is inherently a problem for any firmware. Only an abundance of tiny line segments should wreak havoc, but as long as they are connected to a long segment it shouldn't cause buffer underruns and so it shouldn't be a problem.

I don't think `simplify` should be made to guarantee that no small line segments exist. The function is called all over the place in the engine; if it's a gcode problem then maybe we need a separate filtering step at the gcode generation phase. For example, alter `LayerPlan::addExtrusionMove` to only do anything if the line segment is long enough.

Reply:
Hmm. Actually...

We are in the case where `if (next_length2 > smallest_line_segment_squared)` so there is no escalation problem.

Still I think it's better to adhere to the criterion that no segment longer than `shortest_line_segment` may have it's direction altered, which is possible with option C or D.

Reply:
In a live talk we decided that the effort and code complexity wasnt worth it.